sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala

```diff
@@ -878,6 +898,7 @@ object JdbcUtils extends Logging with SQLConfHelper {
       df: DataFrame,
       tableSchema: Option[StructType],
       isCaseSensitive: Boolean,
+      upsert: Boolean,
```

Comment:
Shall we avoid add parameter?
Please reuse `JdbcOptionsInWrite`.

Reply:
This is needed because `saveTable` is called for all save modes, but only in `Append` mode for an existing table, we want to use the upsert statement, for all other code paths, we want to use plain insert.

We could decrease code complexity by removing this `upsert` argument and use `options.isUpsert`, but that would use upsert statements in situations where no upserts are needed.

Reply:
How do I know in `saveTable` the current save mode?

Reply:
get from `options`

Reply:
The save mode is not part of `options`, it is an argument to `createRelation`.

Reply:
Why the parameter name is `upsert`?

Reply:
It is called `upsert`, because it indicates `saveTable` to use upsert statements, rather than insert statements:

```scala
val insertStmt = if (upsert) {
  getUpsertStatement(table, rddSchema, tableSchema, isCaseSensitive, dialect, options)
} else {
  getInsertStatement(table, rddSchema, tableSchema, isCaseSensitive, dialect)
}
```
I am open to introduce the `Upsert` save mode, but would like to hear other commiters' thoughts on this before I go and add this.

Reply:
For better abstract, I think we should introduce the `Upsert` mode as you said. Some data sources doesn't support the upsert statement supported by some JDBC database, but we still could implement the `Upsert` mode by composite some operations like `Overwrite` mode.